### DISCUSSION TITLE
Add download speed limit support

### DISCUSSION THREAD

**Discussion by shichanglin5** [seaweedfs/seaweedfs#3403] [Ideas]
## Describe
When downloading large files, chunk files are frequently read from the volume, which occupies a lot of disk IO at that time; if the user performs parallel large file downloads, the performance of the entire cluster will become uncontrollable, so I think it is necessary to add a file download speed limit avoid this problem to a large extent;

## Some Cases
Throughput test download file size is 4MB
The number of test virtual users (threads) is 150 (3*50)
Test duration 30 minutes

Throughput test, two "craters" are caused by executing large file downloads (9 file downloads with a size of 5GB are opened at the same time)
<img width="1451" alt="image" src="https://user-images.githubusercontent.com/48707349/182827049-e5960d73-1907-4de5-aefb-e6ff49aebfbd.png">
Cpu & Net Traffic
<img width="726" alt="image" src="https://user-images.githubusercontent.com/48707349/182827755-ca869b63-d049-48e8-8f7e-2107f37e8ce1.png">
Disk
<img width="716" alt="image" src="https://user-images.githubusercontent.com/48707349/182831061-4543aa1b-3b1c-4f7c-b753-69e09a988891.png">

**Comment by shichanglin5** [comment]
@chrislusf What do you think, if there is no problem, I will deal with it

**Reply by shichanglin5** [reply]
![image](https://user-images.githubusercontent.com/48707349/182835550-9504aedc-57b0-4caa-ae61-25578620fb22.png)

**Reply by chrislusf** [reply]
- What is the problem? You want to optimize for fairness/throughput/latency/...?
- Where is the problem? The volume server or the filer? Memory or cpu?
- Why the fix is the best?

**Reply by shichanglin5** [reply]
@chrislusf  Sorry for not making it clear

The problem is that large files will frequently request volumeServer to download chunks, which will affect other requests download, resulting in a significant drop in throughput.

When the filer processes the download request, it will find all the chunks of the file, and then traverse the download; what I want to do is to modify this code to slow down the download frequency of chunks, in this way, to avoid the problem of throughput drop caused by a large number of chunk downloads for a single request , to ensure that the download speed of each request does not exceed the configured threshold

**Reply by chrislusf** [reply]
how many volume servers you used?

**Reply by shichanglin5** [reply]
This problem was found in the production environment. There are 36 volume servers in the production environment. For the convenience of reproduction, only one volume server is used here.

**Comment by kmlebedev** [comment]
@shichanglin5 Have you tried setting it up? ?
```
  -concurrentDownloadLimitMB int
    	limit total concurrent download size (default 256)
  -concurrentUploadLimitMB int
    	limit total concurrent upload size (default 256)
```

**Reply by shichanglin5** [reply]
These two are different. '-concurrentDownloadLimitMB' is the parameter of volumeServer (since the chunk size is fixed, the default value is 4, so concurrent chunk downloads can be limited). 

When the concurrent download exceeds 'concurrentDownloadLimitMB', subsequent requests for the volume will wait. Avoid volumeServer problems due to overload; but for large file downloads, such as a 4g file, the default is divided into 4m chunks, which will be divided into 1000 chunks, while other ordinary file downloads may only be 1 chunk, if there is no restriction, then 1000 chunk download requests may keep occupying disk io (although 'chunkViews' downloads are serial, disk access is also serial), so other chunk download requests may be blocked during large file downloads, so there will be things like As seen in the first screenshot, the throughput test appears "craters". 

During this process, since 1000 chunks are downloaded serially, even if 'concurrentDownloadLimitMB' is not triggered, some ordinary requests may take up io due to large file downloads. The situation where chunk download takes a relatively long time

--- The above is my understanding, if there is anything wrong, please correct me in time @kmlebedev
